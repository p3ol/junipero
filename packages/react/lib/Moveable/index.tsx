import {
  type ComponentPropsWithoutRef,
  type ReactElement,
  type MouseEvent as ReactMouseEvent,
  type Ref,
  cloneElement,
  useCallback,
  useEffect,
  useReducer,
} from 'react';
import { mockState, classNames } from '@junipero/core';

export declare interface MoveableState {
  moving: boolean;
  deltaX: number;
  deltaY: number;
  mouseX: number;
  mouseY: number;
  startX: number;
  startY: number;
}

export declare type MoveableStrategy = 'transform' | 'position';

export declare type MoveableRef = any;

export declare interface MoveableProps extends ComponentPropsWithoutRef<any> {
  ref?: Ref<MoveableRef>;
  x?: number;
  y?: number;
  transformScale?: number;
  strategy?: MoveableStrategy;
  onMouseDown?: (
    e: ReactMouseEvent<HTMLDivElement>
  ) => void;
  onMoveStart?: (
    e: ReactMouseEvent<HTMLDivElement>
  ) => void;
  onMove?: (
    state: MoveableState,
    e: MouseEvent
  ) => void;
  onMoveEnd?: (
    state: MoveableState,
    e: MouseEvent
  ) => void;
}

const Moveable = ({
  ref,
  x,
  y,
  style,
  className,
  children,
  transformScale = 1,
  strategy = 'transform',
  onMouseDown,
  onMoveStart,
  onMove,
  onMoveEnd,
  ...rest
}: MoveableProps) => {
  const [state, dispatch] = useReducer(mockState<MoveableState>, {
    moving: false,
    mouseX: 0,
    mouseY: 0,
    deltaX: (x || 0),
    deltaY: (y || 0),
    startX: 0,
    startY: 0,
  });

  const onMouseDown_ = useCallback((e: ReactMouseEvent<HTMLDivElement>) => {
    dispatch({
      moving: true,
      mouseX: e.clientX,
      mouseY: e.clientY,
      startX: state.deltaX,
      startY: state.deltaY,
    });

    onMoveStart?.(e);
    onMouseDown?.(e);
  }, [state.deltaX, state.deltaY, onMoveStart, onMouseDown]);

  const onMouseMove = useCallback((e: MouseEvent) => {
    if (!state.moving) return;

    state.deltaX = (e.clientX - state.mouseX +
      (state.startX * transformScale)) / transformScale;
    state.deltaY = (e.clientY - state.mouseY +
      (state.startY * transformScale)) / transformScale;

    dispatch({ deltaX: state.deltaX, deltaY: state.deltaY });
    onMove?.(state, e);
  }, [state, onMove, transformScale]);

  const onMouseUp = useCallback((e: MouseEvent) => {
    state.moving = false;
    state.mouseX = 0;
    state.mouseY = 0;
    state.startX = state.deltaX;
    state.startY = state.deltaY;

    dispatch({
      moving: false,
      mouseX: 0,
      mouseY: 0,
      startX: state.deltaX,
      startY: state.deltaY,
    });

    onMoveEnd?.(state, e);
  }, [state, onMoveEnd]);

  useEffect(() => {
    if (state.moving) {
      window.addEventListener('mousemove', onMouseMove);
      window.addEventListener('mouseup', onMouseUp);
    } else {
      window.removeEventListener('mousemove', onMouseMove);
      window.removeEventListener('mouseup', onMouseUp);
    }

    return () => {
      window.removeEventListener('mousemove', onMouseMove);
      window.removeEventListener('mouseup', onMouseUp);
    };
  }, [state.moving, onMouseMove, onMouseUp]);

  const child: ReactElement<
      ComponentPropsWithoutRef<any>
    > = typeof children !== 'string' && Array.isArray(children)
      ? children[0] : children;

  return cloneElement(child, {
    ...rest,
    ref,
    className: classNames(
      className,
      child.props?.className,
      'junipero moveable',
      {
        moving: state.moving,
      }
    ),
    style: {
      ...style,
      ...strategy === 'position' ? {
        position: 'relative',
        left: `${state.deltaX}px`,
        top: `${state.deltaY}px`,
      } : {
        transform: (style?.transform || '') +
          ` translate3d(${state.deltaX}px, ${state.deltaY}px, 0)`,
      },
    },
    onMouseDown: onMouseDown_,
  } as ComponentPropsWithoutRef<any>);
};

Moveable.displayName = 'Moveable';

export default Moveable;
