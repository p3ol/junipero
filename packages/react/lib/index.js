// Display
export { default as Abstract } from './Abstract';
export { default as Alert } from './Alert';
export { default as Alerts } from './Alerts';
export { default as AlertsControl } from './AlertsControl';
export { default as Badge } from './Badge';
export { default as BreadCrumb } from './BreadCrumb';
export { default as BreadCrumbItem } from './BreadCrumbItem';
export { default as Button } from './Button';
export { default as Calendar } from './Calendar';
export { default as Card } from './Card';
export { default as Draggable } from './Draggable';
export { default as Dropdown } from './Dropdown';
export { default as DropdownGroup } from './DropdownGroup';
export { default as DropdownItem } from './DropdownItem';
export { default as DropdownMenu } from './DropdownMenu';
export { default as DropdownToggle } from './DropdownToggle';
export { default as Droppable } from './Droppable';
export { default as Label } from './Label';
export { default as List } from './List';
export { default as ListCell } from './ListCell';
export { default as ListColumn } from './ListColumn';
export { default as ListItem } from './ListItem';
export { default as Modal } from './Modal';
export { default as Slider } from './Slider';
export { default as Spinner } from './Spinner';
export { default as Tab } from './Tab';
export { default as Tabs } from './Tabs';
export { default as Tag } from './Tag';
export { default as Toast } from './Toast';
export { default as Toasts } from './Toasts';
export { default as ToastsControl } from './ToastsControl';
export { default as Toggle } from './Toggle';
export { default as Tooltip } from './Tooltip';
export { default as TouchableZone } from './TouchableZone';
export { default as Transition } from './Transition';

// Forms
export { default as CheckboxField } from './CheckboxField';
export { default as CodeField } from './CodeField';
export { default as ColorField } from './ColorField';
export { default as DateField } from './DateField';
export { default as FieldAddon } from './FieldAddon';
export { default as FieldControl } from './FieldControl';
export { default as FieldGroup } from './FieldGroup';
export { default as RadioField } from './RadioField';
export { default as SelectField } from './SelectField';
export { default as TextField } from './TextField';

// Logos
export {
  PooolLogo,
  PooolCompactLogo,
  AccessLogo,
  SubscribeLogo,
  ConnectLogo,
  EngageLogo,
  FlowLogo,
} from './Logo';

// Core
export {
  classNames,
  addClass,
  removeClass,
  mockState,
  isUndefined,
  isNull,
  isArray,
  isObject,
  isDate,
  exists,
  get,
  set,
  omitBy,
  omit,
  pick,
  cloneDeep,
  fromPairs,
  mergeDeep,
  ensureNode,
  findDeep,
  filterDeep,
  ensureMinMax,
  getFloatPrecision,
  formatThousand,
} from '@junipero/core';

// Hooks
export {
  useEventListener,
  useInterval,
  useTimeout,
  useEffectAfterMount,
  useLayoutEffectAfterMount,
} from '@junipero/hooks';

export {
  useAlerts,
  useToasts,
} from './hooks';

// Icons
export {
  Arrows,
  Remove,
  ArrowLeft,
  ArrowRight,
  ArrowDown,
  ArrowUp,
  Time,
  Check,
} from './icons';
