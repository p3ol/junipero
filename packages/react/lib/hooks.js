import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
