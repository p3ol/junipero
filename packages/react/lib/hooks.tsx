import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
