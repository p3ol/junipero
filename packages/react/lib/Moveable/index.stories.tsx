import { useState } from 'react';

import { useInfiniteCanvas } from '../hooks';
import Button from '../Button';
import InfiniteCanvas from '../InfiniteCanvas';
import Moveable, { type MoveableProps } from '.';

export default { title: 'react/Moveable' };

export const Basic = () => (
  <Moveable>
    <Button>Click me!</Button>
  </Moveable>
);

export const WithPositionStrategy = () => (
  <Moveable strategy="position">
    <Button>Click me!</Button>
  </Moveable>
);

const ZoomedMoveable = (props: MoveableProps) => {
  const { zoom } = useInfiniteCanvas();

  return (
    <Moveable { ...props } transformScale={zoom} />
  );
};

export const InsideInfiniteCanvas = () => (
  <InfiniteCanvas
    className="fixed top-0 left-0 w-screen h-screen"
  >
    <ZoomedMoveable>
      <Button>Click me!</Button>
    </ZoomedMoveable>
  </InfiniteCanvas>
);

export const Controlled = () => {
  const [position, setPosition] = useState({ x: 0, y: 0 });
  const [disabled, setDisabled] = useState(false);

  return (
    <div>
      <Button
        onClick={() => setPosition({ x: position.x + 10, y: position.y + 10 })}
      >
        Move elsewhere
      </Button>
      <Button
        onClick={() => setDisabled(!disabled)}
        className="ml-2"
      >
        {disabled ? 'Enable' : 'Disable'}
      </Button>
      <Moveable
        x={position.x}
        y={position.y}
        disabled={disabled}
        onMoveEnd={state => setPosition({ x: state.deltaX, y: state.deltaY })}
      >
        <Button>Click me!</Button>
      </Moveable>
    </div>
  );
};
