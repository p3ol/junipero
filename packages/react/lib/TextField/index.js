import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
