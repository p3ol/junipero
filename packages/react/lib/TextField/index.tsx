import {
  type ChangeEvent,
  type FocusEvent,
  type RefObject,
  type WheelEvent,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, exists } from '@junipero/core';

import type {
  FieldContent,
  JuniperoRef,
  SpecialComponentPropsWithRef,
} from '../types';
import { mockState } from '../utils';
import { useFieldControl } from '../hooks';

export declare interface TextFieldRef extends JuniperoRef {
  dirty: boolean;
  focused: boolean;
  valid: boolean;
  value: string | number;
  blur(): void;
  focus(): void;
  reset(): void;
  setDirty(dirty: boolean): void;
  innerRef: RefObject<HTMLDivElement>;
  inputRef: RefObject<HTMLInputElement | HTMLTextAreaElement>;
}

export declare interface TextFieldProps
  extends SpecialComponentPropsWithRef<any, TextFieldRef> {
  autoFocus?: boolean;
  disabled?: boolean;
  required?: boolean;
  type?: string;
  valid?: boolean;
  value?: string | number;
  onBlur?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onChange?(field: FieldContent<string | number>): void;
  onFocus?(event: FocusEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onWheel?(event: WheelEvent<HTMLInputElement | HTMLTextAreaElement>): void;
  onValidate?(
    val: string | number,
    flags: { required: boolean; dirty: boolean }
  ): boolean;
}

export declare interface TextFieldState {
  value: string | number;
  valid: boolean;
  dirty: boolean;
  focused: boolean;
}

const TextField = ({
  ref,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val: string | number, opts?: { required: boolean }) =>
    !!val || !opts?.required,
  ...rest
}: TextFieldProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const inputRef = useRef<HTMLInputElement & HTMLTextAreaElement>(null);
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState<TextFieldState>, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = (value?: string) => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = (
    e: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = (e: FocusEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = (dirty: boolean) => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? ((_: WheelEvent) => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
};

TextField.displayName = 'TextField';

export default TextField;
