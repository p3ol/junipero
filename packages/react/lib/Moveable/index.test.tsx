import { fireEvent, render } from '@testing-library/react';

import Moveable from '.';

describe('<Moveable />', () => {
  it('should render a moveable', () => {
    const { container, unmount } = render(
      <Moveable>
        <div style={{ width: '1000px', height: '1000px' }}>Content</div>
      </Moveable>
    );

    expect(container).toMatchSnapshot();
    fireEvent.mouseDown(container.querySelector('.moveable'), { button: 0 });

    expect(container).toMatchSnapshot('moving');

    fireEvent.mouseMove(container.querySelector('.moveable'), {
      clientX: 100,
      clientY: 100,
    });
    fireEvent.mouseUp(container.querySelector('.moveable'), {
      clientX: 100,
      clientY: 100,
    });
    expect(container).toMatchSnapshot('moveable moved');
    unmount();
  });

  it('should not move when disabled', () => {
    const { container, unmount } = render(
      <Moveable disabled>
        <div style={{ width: '1000px', height: '1000px' }}>Content</div>
      </Moveable>
    );

    expect(container).toMatchSnapshot();
    fireEvent.mouseDown(container.querySelector('.moveable'), { button: 0 });

    expect(container).toMatchSnapshot('not moving');

    fireEvent.mouseMove(container.querySelector('.moveable'), {
      clientX: 100,
      clientY: 100,
    });
    fireEvent.mouseUp(container.querySelector('.moveable'), {
      clientX: 100,
      clientY: 100,
    });
    expect(container).toMatchSnapshot('still not moving');
    unmount();
  });

  it('should move on a grid', () => {
    const { container, unmount } = render(
      <Moveable step={50}>
        <div style={{ width: '1000px', height: '1000px' }}>Content</div>
      </Moveable>
    );

    expect(container).toMatchSnapshot();
    fireEvent.mouseDown(container.querySelector('.moveable'), { button: 0 });

    expect(container).toMatchSnapshot('moving');

    fireEvent.mouseMove(container.querySelector('.moveable'), {
      clientX: 80,
      clientY: 80,
    });
    fireEvent.mouseUp(container.querySelector('.moveable'), {
      clientX: 80,
      clientY: 80,
    });
    expect(container).toMatchSnapshot('moveable moved to 100,100');
    unmount();
  });
});
