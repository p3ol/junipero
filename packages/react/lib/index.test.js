import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
