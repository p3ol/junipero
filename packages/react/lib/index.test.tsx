import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
