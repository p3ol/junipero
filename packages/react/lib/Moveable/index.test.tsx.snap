// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Moveable /> > should render a moveable > moveable moved 1`] = `
<div>
  <div
    class="junipero moveable"
    style="transform: translate3d(100px, 100px, 0);"
  >
    Content
  </div>
</div>
`;

exports[`<Moveable /> > should render a moveable > moving 1`] = `
<div>
  <div
    class="junipero moveable moving"
    style="transform: translate3d(0px, 0px, 0);"
  >
    Content
  </div>
</div>
`;

exports[`<Moveable /> > should render a moveable 1`] = `
<div>
  <div
    class="junipero moveable"
    style="transform: translate3d(0px, 0px, 0);"
  >
    Content
  </div>
</div>
`;
